backend/mr-eventing/src/main/kotlin/io/featurehub/mr/events/converter/FeatureMessagingConverterImpl.kt

```diff
@@ -0,0 +1,106 @@
+package io.featurehub.mr.events.converter
+
+import io.featurehub.db.api.RolloutStrategyUpdate
+import io.featurehub.messaging.model.*
+import io.featurehub.mr.events.common.converter.FeatureMessagingConverter
+import io.featurehub.mr.events.common.converter.FeatureMessagingParameter
+import io.featurehub.mr.model.RolloutStrategy
+import io.featurehub.mr.model.RolloutStrategyAttribute
+import java.time.ZoneOffset
+import java.util.ArrayList
+
+class FeatureMessagingConverterImpl : FeatureMessagingConverter{
+  override fun toFeatureMessagingUpdate(
+    featureMessagingParameter: FeatureMessagingParameter
+  ): FeatureMessagingUpdate {
+
+    val featureValue = featureMessagingParameter.featureValue
+    val environment = featureValue.environment
+    val parentApplication = environment.parentApplication
+    val portfolio = parentApplication.portfolio
+    return FeatureMessagingUpdate()
+      .featureKey(featureValue.feature.key)
+      .environmentId(environment.id)
+      .whoUpdated(featureValue.whoUpdated.name)
+      .whenUpdated(featureValue.whenUpdated.atOffset(ZoneOffset.UTC))
+      .applicationId(parentApplication.id)
+      .portfolioId(portfolio.id)
+      .organizationId(portfolio.organization.id)
+      .let {
+        val defaultValueUpdate = featureMessagingParameter.defaultValueUpdate
+        val defaultValueUpdated = defaultValueUpdate?.updated
+        val defaultValuePrevious = defaultValueUpdate?.previous
+        if (defaultValueUpdate?.hasChanged == true && defaultValueUpdated != null && defaultValuePrevious != null) it.featureValueUpdated(
+          MessagingFeatureValueUpdate()
+            .valueType(featureValue.feature.valueType)
+            .updated(defaultValueUpdated)
+            .previous(defaultValuePrevious)
+        )
+        else it
+      }
+      .let {
+        val lockUpdated = featureMessagingParameter.lockUpdate?.updated
+        val lockPrevious = featureMessagingParameter.lockUpdate?.previous
+        if (featureMessagingParameter.lockUpdate?.hasChanged == true && lockUpdated != null && lockPrevious != null)
```

Comment:
regarding this and retired in relation to nullable, the only reason a lock would be null is if it went from not existing to existing or back again. For non-boolean values, we assume the value is always unlocked (null = !locked, i.e. false) so you may wish to backtrack a bit on this object (same with retired, null = false) and assume a null is always false, which might simplify this logic?